sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                // means a failure has happened, so try closing it and starting a new one.  In case we don't have a task that should
+                // exist, create it.  This might happen if the user hasn't updated ActivePartitionProcessors when initializing processing
+                // in the previous cycle.
+
+                await Task.WhenAll(InstanceOwnership
+                    .Select(async kvp =>
+                    {
+                        if (!ActivePartitionProcessors.TryGetValue(kvp.Key, out Task processingTask) || processingTask.IsCompleted)
+                        {
+                            await StopPartitionProcessingIfRunningAsync(kvp.Key, ProcessingStoppedReason.Exception).ConfigureAwait(false);
+
+                            var context = CreateContext(kvp.Key);
+                            await InitializeProcessingForPartitionAsync(context).ConfigureAwait(false);
+                        }
+                    }))
+                    .ConfigureAwait(false);
+
+                // Get a complete list of the partition ids present in the Event Hub.  This should be immutable for the time being, but
+                // it may change in the future.
+
+                var partitionIds = await connection.GetPartitionIdsAsync(RetryPolicy).ConfigureAwait(false);
+
+                // Find an ownership to claim and try to claim it.  The method will return null if this instance was not eligible to
+                // increase its ownership list, if no claimable ownership could be found or if a claim attempt has failed.
+
+                var claimedOwnership = await FindAndClaimOwnershipAsync(partitionIds, completeOwnershipList, activeOwnership).ConfigureAwait(false);
+
+                if (claimedOwnership != null)
+                {
+                    InstanceOwnership[claimedOwnership.PartitionId] = claimedOwnership;
+
+                    var context = CreateContext(claimedOwnership.PartitionId);
+                    await InitializeProcessingForPartitionAsync(context).ConfigureAwait(false);
+                }
+
+                // Wait the remaining time, if any, to start the next cycle.  The total time of a cycle defaults to 10 seconds,
+                // but it may be overridden by a derived class.
+
+                TimeSpan remainingTimeUntilNextCycle = LoadBalanceUpdate - cycleDuration.Elapsed;
+
+                if (remainingTimeUntilNextCycle > TimeSpan.Zero)
+                {
+                    // If a stop request has been issued, Task.Delay will throw a TaskCanceledException.  This is expected and it
+                    // will be caught by the StopAsync method.
+
+                    await Task.Delay(remainingTimeUntilNextCycle, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // If cancellation has been requested, throw an exception so we can keep a consistent behavior.
+
+            if (cancellationToken.IsCancellationRequested)
```

Comment:
There's a nice little helper extension for `ThrowIfCanceled<TaskCanceledException>` in our `Core` namespace.